docs/ReceiveBuffer.md

```diff
@@ -0,0 +1,223 @@
+The AppOwned mode is the exception, with memory being managed by the caller. 
+
+Internally, the receive buffer tries to minimize total memory usage, both in terms of bytes and number of unique buffers to keep at once.
+It generally prefers one large chunk over multiple smaller chunks, even if this means data must be copied from one buffer to another (larger) one.
+This aims to reduce the complexity of both the internal implementation and the caller implementation.
+
+Since the receive buffer gives out pointers to its internal buffers, it must keep track of these external references and
+extend the lifetimes of chunks so that they are not deleted from underneath the caller.
+This means the receive buffer might keep a "retired" buffer for some time, or decide to have multiple active chunks for a while.
+
+Another aspect of memory management is balancing the amount of memory we advertised to the peer we are willing to allocate and how much we actually allocate.
+The caller controls both the advertised, maximum allocation size, as well as the initial buffer size to allocate.
+The caller can dynamically increase the maximum size as necessary.
+
+The receive buffer takes these values and dynamically allocates memory (doubling in size as necessary) up to the maximum
+size - except in AppOwned mode, where the application must provide memory to match the maximum size advertised to the peer.
+
+### Less words, more ASCII art!!
+
+- `x`: A 'write' put data here
+- `.`: No data here yet
+- ` `: Bytes that should no longer be accessed
+
+```
+   Retired                                         Chunk 1                            Chunk 2              Chunk 3       
+┌──────────────┐   ┌───────────────────────────────────────────────────────────┐   ┌────────────────┐   ┌──────────────┐ 
+│              │   │xxx.......xxxxxxxx                   xxxxxxxxxxxxxxxxxxxxxx├──►│xxxxxxxx........├──►│xxxxxxxxxxx...│ 
+└──────────────┘   └───▲──────────────▲──────────────────▲───────────────▲─────┘   └────────────────┘   └──────────────┘ 
+                       │              │                  │               │                                               
+                   ReadStart +      ReadStart +       ReadStart       ReadStart +                                        
+                    ReadLength       Capacity                          ReadPendingLength                                 
+```
+
+This receive buffer has 1 retired chunk and 3 active chunks.
+`WrittenRanges` contains 3 data segments:
+- `[0, ReadStart + ReadLength]`, cycling through the end of Chunk 1;
+- The second segment goes over the end of Chunk 1 and the start of Chunk 2;
+- The third segment is at the start of chunk3
+
+**Remark**: The state of the receive buffer presented above would actually never happen in practice - it is so generic it violates each mode constraints.
```

Comment:
This is great documentation - my only nit is that instead of providing one sample of illegal state, it would be better to provide several valid examples covering the same sets of scenarios.

Reply:
That's fair. I want the "illegal" state, because I find it a great way to frame reasoning about this code (I am planning some simplifications based on it).

I though about adding "per-mode" examples, or a "timeline" to show how operations affect it, but I grew lazy. I'll try to add something so we don't only have the "invalid" case.

Reply:
Thanks, this is exactly what I had in mind